### PR TITLE
Feature/ad integration

### PR DESCRIPTION
**Pull request by lzammit**
Support to unreg node if an Active directory account is disabled or locked.

### CODE REVIEW

File: addons/AD/unreg_node_deleted_account.ps1 [hunk 1]
```diff
@@ -0,0 +1,25 @@
+########################################################################################
+#Powershell script to unregister locked Active Directory account based on the UserName.#
+########################################################################################
+
+Get-EventLog -LogName Security -InstanceId 4725 |
+   Select ReplacementStrings,"Account name"|
+   % {
+    $url = "https://@IP_PACKETFENCE:9090/"
+    [System.Net.ServicePointManager]::ServerCertificateValidationCallback = {$true}
```
Comment (fdurand) [inline addons/AD/unreg_node_deleted_account.ps1:11]: Somewhere in the script you have to define the webservice uername and password.

File: addons/AD/unreg_node_locked_account.ps1 [hunk 2]
```diff
@@ -0,0 +1,26 @@
+#########################################################################################
+#Powershell script to unregister locked Active Directory account based on the UserName.#
+#########################################################################################
+
+Get-EventLog -LogName Security -InstanceId 4725 |
+   Select ReplacementStrings,"Account name"|
+   % {
+    $url = "https://@IP_PACKETFENCE:9090/"
+    [System.Net.ServicePointManager]::ServerCertificateValidationCallback = {$true}
```
Comment (fdurand) [inline addons/AD/unreg_node_locked_account.ps1:11]: Idem username and password

**Comment by fdurand** [comment]
NEWS file ?

File: lib/pf/api.pm [hunk 3]
```diff
@@ -98,6 +99,24 @@ sub update_iplog {
     return (pf::iplog::iplog_update($srcmac, $srcip, $lease_length));
 }
  
+sub unreg_node_for_pid {
+    my ($class, $pid) = @_;
+
+    my $logger = pf::log::get_logger();
+
+    #use Data::Dumper;
+    #$logger->warn(Dumper $pid);
```
Comment (fdurand) [inline lib/pf/api.pm]: Useless

File: lib/pf/node.pm [hunk 4]
```diff
@@ -356,6 +363,16 @@ sub node_pid {
 }
 
 #
+# return mac for specified register pid
+#
+sub node_view_reg_pid {
+    my ($pid) = @_;
+#    my $query = db_query_execute(NODE, $node_statements, 'node_pid_sql', $pid);
+#    my ($count) = $query->fetchrow_array();
```
Comment (fdurand) [inline lib/pf/node.pm]: useless

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 5]
```diff
@@ -2376,6 +2376,161 @@ realm ntdomain {
 }
 ----
 
+AD-Integration:
+~~~~~~~~~~~~~~~
+
+Deleted Account:
+++++++++++++++++
+
+Create a file unreg_node_deleted_account.ps1 on the Windows Server and make sure to change the @IP_PACKETFENCE. I am using username and password "admin" for the werbservice credential, make sure they match under Configuration > Web services.
```
Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc]: **for the web services credentials

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 6]
```diff
@@ -2376,6 +2376,161 @@ realm ntdomain {
...
+    $stream.close()
+
+    $reader = New-Object System.IO.Streamreader -ArgumentList $web.GetResponse().GetResponseStream()
+    $reader.ReadToEnd()
+    $reader.Close()
+ 
+   }
+----
+
+Create the scheduled task based on event ID:
```
Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc]: **based on an event ID

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 7]
```diff
@@ -2376,6 +2376,161 @@ realm ntdomain {
...
+----
+Action: Start a program
+Program/script: powershell.exe
+Add arguments (optional): C:\scripts\unreg_node_locked_account.ps1
+----
+
+Settings:
+
+----
+At the bottom, select in the list "Run a new instance in parallel"
```
Comment (julsemaan) [inline docs/PacketFence_Administration_Guide.asciidoc:2604]: Maybe say the what option this list refers to ?

**Comment by fdurand** [comment]
Can be merge for me

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 8]
```diff
@@ -2376,6 +2376,161 @@ realm ntdomain {
 }
 ----
 
+AD-Integration:
+~~~~~~~~~~~~~~~
+
+Deleted Account:
+++++++++++++++++
+
+Create a file unreg_node_deleted_account.ps1 on the Windows Server and make sure to change the @IP_PACKETFENCE. I am using username and password "admin" for the werb services credentials, make sure they match under Configuration > Web services.
```
Comment (cgx) [inline docs/PacketFence_Administration_Guide.asciidoc]: Suggested rewording: _Make sure the username and password match the credentials defined in the Web admin interface under Configuration > Web Services._

File: docs/PacketFence_Administration_Guide.asciidoc [hunk 9]
```diff
@@ -2376,6 +2376,161 @@ realm ntdomain {
...
+----
+At the bottom, select in the list "Run a new instance in parallel" in order to unregister multiple nodes at the same time.
+----
+
+Validate with Ok and give the account who will run this task. (Usually DOMAIN\Administrator)
+
+Locked Account:
+++++++++++++++++
+
+Create a file unreg_node_locked_account.ps1 on the Windows Server and make sure to change the @IP_PACKETFENCE. I am using username and password "admin" for the web services credentials, make sure they match under Configuration > Web services.
```
Comment (cgx) [inline docs/PacketFence_Administration_Guide.asciidoc]: Suggested rewording: _Make sure the username and password match the credentials defined in the Web admin interface under Configuration > Web Services._

File: lib/pf/api.pm [hunk 10]
```diff
@@ -98,6 +99,22 @@ sub update_iplog {
     return (pf::iplog::iplog_update($srcmac, $srcip, $lease_length));
 }
  
+sub unreg_node_for_pid {
+    my ($class, $pid) = @_;
+
+    my $logger = pf::log::get_logger();
+
+    $logger->info("Unregistering nodes for $pid");
```
Comment (cgx) [inline lib/pf/api.pm]: Move this line after fetching the `@node_infos` and log the number of nodes we're going to unregister. Example:

`$logger->info("Unregistering ".scalar(@node_infos)." nodes for $pid");`

File: lib/pf/node.pm [hunk 11]
```diff
@@ -191,6 +192,12 @@ sub node_db_prepare {
         WHERE node.mac=?
     SQL
 
+    $node_statements->{'node_view_reg_pid_sql'} = get_db_handle()->prepare(<<'    SQL');
+        SELECT node.mac
+        FROM node
+        WHERE node.pid=? AND node.status="reg";
```
Comment (cgx) [inline lib/pf/node.pm]: Use the constant `$STATUS_REGISTERED` instead of `reg`.

**Pull request merged**